### PR TITLE
`BorderRadiusControl`: Stop using `UnitControl`’s deprecated `unit` prop

### PR DESCRIPTION
**Pull request by stokesman**
## What?
Related to #39503 , this PR removes the use of `UnitControl`’s deprecated `unit` prop in `BorderRadiusControl`.

## Why?
The `unit` prop is marked as deprecated, the component's docs recommend passing the unit directly through the `value` prop

## How?
Remove the `unit` prop on sub-component `AllInputControl`. It is redundant since the values already contain the unit.

## Testing Instructions
- Tests pass
- Component behaves as expected in the block editor:
    1. Create a new page/post
    2. Add a Group block
    3. Play around with the border radius controls in the sidebar.
    4. Unlink the sides and change at least one of the sides to have a different unit and value*.
    5. Link the sides, type a number and verify the unit defaults to "px".

*I think there is a bug in this control. When the sides are unlinked and units vary but the radius values are all the same number, switching back to linked sides does not show the "Mixed" placeholder in the input.

### CODE REVIEW

**Review by ciampo — APPROVED** [review]
I was literally about to start working on this task, so extra thanks for opening this PR, @stokesman !

The code changes here LGTM and they test well as per instructions 🚀 

> I think there is a bug in this control. When the sides are unlinked and units vary but the radius values are all the same number, switching back to linked sides does not show the "Mixed" placeholder in the input.

I can confirm this is happening also for me.

I also came across another unexpected behavior while playing around this this component (in my case, in the global styles sidebar):
  - I edited the border width and radius for the columns block, saved the global styles and reloaded the page
  - If I delete the values for border radius (and width) in the `UnitControl` inputs, the borders don't update as if I had removed them — to achieve that, I have to explicitly input `0`. The same behavior happens when "resetting" this fields through the `ToolsPanel` dropdown.
  - Basically, I expected that deleting the value in these input would be equivalent to "delete" those values, instead of resetting them to the last saved status

https://user-images.githubusercontent.com/1083581/158855082-86e60607-34ae-4cf6-9cb0-dbe6c0bc16f6.mp4

Just pinging @aaronrobertshaw and @andrewserong who might be able to confirm if the two behaviors listed above are unintended — in any ways, they should be treated separately from this PR.

**Comment by stokesman** [comment]
>I was literally about to start working on this task,

I'm glad you saw this PR first then! The root reason I made this is because I noticed #39531 didn't fix that issue for `BorderRadiusControl`. Thanks to having seen your PRs, I thought maybe the `unit` prop  was interfering and that does seem to be the case.

**Comment by andrewserong** [comment]
> Just pinging @aaronrobertshaw and @andrewserong who might be able to confirm if the two behaviors listed above are unintended — in any ways, they should be treated separately from this PR.

Thanks for the ping!

> If I delete the values for border radius (and width) in the UnitControl inputs, the borders don't update as if I had removed them — to achieve that, I have to explicitly input 0. The same behavior happens when "resetting" this fields through the ToolsPanel dropdown.
> Basically, I expected that deleting the value in these input would be equivalent to "delete" those values, instead of resetting them to the last saved status

I think this also happens with `padding` too — if, after removing the values, I then save the global styles and reload the page though, the removed values persist. So it looks like it's mostly a UI issue / to do with updating global styles values on the fly (and having those changes reflected in the editor in real-time), rather than that it's not possible to clear the values altogether?

> in any ways, they should be treated separately from this PR.

Yes, totally something for us to investigate separately 🙂

**Comment by andrewserong** [comment]
> I think there is a bug in this control. When the sides are unlinked and units vary but the radius values are all the same number, switching back to linked sides does not show the "Mixed" placeholder in the input.

> I can confirm this is happening also for me.

I think this is because [getAllValue](https://github.com/wordpress/gutenberg/blob/0ea49d674a376636db5c0722ae41083394df9e60/packages/block-editor/src/components/border-radius-control/utils.js#L57) appears to do its best to display the value and uses `mode` so that the unit is the most common unit in the set (so I think it was probably intentional?). An argument could be made here that it's better to show a value to the user rather than `mixed` so that in the collapsed state there's a better indication of the current value if someone wants to change it all at once.

But again, probably a good thing to explore in a separate issue / PR if we think it'd be good to change.

This PR is testing well for me, though!

**Review by andrewserong — APPROVED** [review]

**Comment by aaronrobertshaw** [comment]
> I think this is because [getAllValue](https://github.com/wordpress/gutenberg/blob/0ea49d674a376636db5c0722ae41083394df9e60/packages/block-editor/src/components/border-radius-control/utils.js#L57) appears to do its best to display the value and uses mode so that the unit is the most common unit in the set (so I think it was probably intentional?).

From memory, this followed the `BoxControl` and how it worked.

> An argument could be made here that it's better to show a value to the user rather than mixed so that in the collapsed state there's a better indication of the current value if someone wants to change it all at once.

If the value was shown and the `mode` unit selected it wouldn't indicate that the values were mixed at all would it?

**Review by aaronrobertshaw — APPROVED** [review]
While testing this I think I encountered a possible regression with the UX. I believe previously a user could select a non `px` unit prior to entering a value in the input field. At the moment, switching the unit immediately puts the input field in a `Mixed` state.

I don't think it is related directly to this PR and might have been introduced via recent changes around the `UnitControl` (e.g. https://github.com/WordPress/gutenberg/pull/38987?).

@andrewserong is already on top of it and has a fix up in https://github.com/WordPress/gutenberg/pull/39563.

I'd say this should be right to merge.

**Comment by ciampo** [comment]
Thank you so much for the reviews!

> While testing this I think I encountered a possible regression with the UX [...] might have been introduced via recent changes around `UnitControl` (e.g. https://github.com/WordPress/gutenberg/pull/38987?)
> @andrewserong  is already on top of it and has a fix up in https://github.com/WordPress/gutenberg/pull/39563.

That's some super fast turnaround! Thank you again!


---

Regarding potential follow-ups:

### When the sides are unlinked and units vary but the radius values are all the same number, switching back to linked sides does not show the "Mixed" placeholder in the input.

I confirm that is also how `BoxControl` works. I'm personally not sure about what's the best solution here, both approaches seem reasonable.

(also, I realised how similar the logic of this `BorderRadiusControl` component is to `BoxControl`, any changes we could simplify/share some code across the 2 components? Maybe we could look into extracting this pattern of "linked sides with one control or unlinked sides with as many controls as needed" and reuse it across components? Just a thought!)

### Clearing the quantity from the input causes the last saved/persisted value to be used, instead of acting as "I don't want any padding/border/etc". As a consequence, this happens also when resetting values from `ToolsPanel`.

@andrewserong summarised it quite well:

> So it looks like it's mostly a UI issue / to do with updating global styles values on the fly (and having those changes reflected in the editor in real-time), rather than that it's not possible to clear the values altogether?

**Comment by aaronrobertshaw** [comment]
> I realised how similar the logic of this BorderRadiusControl component is to BoxControl, any changes we could simplify/share some code across the 2 components?

I'd like to think there is some common functionality we could extract. A possible first step might just be to refactor the util functions used given the controls have different icons, slider and layouts for the unlinked controls etc.

**Comment by ciampo** [comment]
Agreed, @aaronrobertshaw !

Regarding this issue:

> So it looks like it's mostly a UI issue / to do with updating global styles values on the fly (and having those changes reflected in the editor in real-time), rather than that it's not possible to clear the values altogether?

Would you or @andrewserong be able to follow up on this one? It's an area of the codebase that I don't know very well

**Comment by aaronrobertshaw** [comment]
> Would you or @andrewserong be able to follow up on this one? It's an area of the codebase that I don't know very well

I'll follow up on this. Hopefully tomorrow.
